src/deepsparse/pipelines/function_pipeline.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Dict, List, Tuple, Type, Union
+
+import numpy
+from pydantic import BaseModel
+
+from deepsparse.pipeline import Pipeline
+from deepsparse.utils.onnx import model_to_path
+
+
+@Pipeline.register(task="custom")
+class FunctionPipeline(Pipeline):
+    """
+    A utility class provided to make specifying custom pipelines easier.
+    Instead of creating a subclass of Pipeline, you can instantiate this directly
+    by passing in functions to call for pre and post processing.
+
+    For example, YOLO pipline could be specified as:
+    ```python
+    def yolo_preprocess(inputs: YOLOInput) -> List[np.ndarray]:
+        ...
+
+    def yolo_postprocess(engine_outputs: List[np.ndarray]) -> YOLOOutput:
+        ...
+
+    yolo = FunctionPipeline(
+        model_path="...",
+        input_schema=YOLOInput,
+        output_schema=YOLOOutput,
+        process_inputs_fn=yolo_preprocess,
+        process_outputs_fn=yolo_postprocess,
+    )
+    ```
+
+    :param model_path: path on local system or SparseZoo stub to load the model from.
+        Passed to :class:`Pipeline`.
+    :param input_schema: A pydantic schema that describes the input to
+        `process_inputs_fn`.
+    :param output_schema: A pydantic schema that describes the output from
+        `process_outputs_fn`.
+    :param process_inputs_fn: A callable (function, method, lambda, etc) that maps
+        an `InputSchema` object to a list of numpy arrays that can be directly passed
+        into the forward pass of the pipeline engine.
+    :param process_outputs_fn: A callable (function, method, lambda, etc) that maps
+        the list of numpy arrays that are the output of the engine forward pass
+        into an `OutputSchema` object.
+    """
+
+    def __init__(
+        self,
+        model_path: str,
+        input_schema: Type[BaseModel],
```

Comment:
What happens if the user passes none here in the event that they don't want to specify an input and output schema?

Reply:
schemas are required for other parts of the pipeline right? not sure what happens now but I can add some type checking in the constructor

Reply:
Yeah, main point for me here was whether or not we could get by with not requiring an input and output schema for the most basic pipelines. I'd like to have it so the base case is someone could create a custom pipeline by simpling passing in the model and it would work. Additions on top of that would then be pre and post processing functions. And then further additions for additional functionality would be the input and output schemas

Reply:
I see the vision yeah that makes sense. I think it'd be straightforward to not require passing the processing functions and just pass input/output to the schemas. It'd probably involve a bunch more changes to not require the schemas - for example the dynamic batch stuff checks various things on the schemas right now, and I know there's at least a little bit of validation done (looks easy to remove though).

I can try out the no functions tomorrow, but the no schemas will probably have to wait until after the release

Reply:
Can we get a basic pipeline working without dynamic batch and anything else enabled?

Reply:
Was pretty easy to do this after all - users now don't need to pass anything other than model path, and they will be interacting directly with engine in this case. see unit tests for example